lib/core/test/unit/Cardano/Wallet/Gen.hs

```diff
@@ -259,3 +272,48 @@ shrinkTxMetadataValue (TxMetaList xs) =
 shrinkTxMetadataValue (TxMetaNumber i) = TxMetaNumber <$> shrink i
 shrinkTxMetadataValue (TxMetaBytes b) = TxMetaBytes <$> shrinkByteString b
 shrinkTxMetadataValue (TxMetaText s) = TxMetaText <$> shrinkText s
+
+instance Arbitrary Natural where
+    shrink = shrinkIntegral
+    arbitrary = arbitrarySizedNatural
+
+genScript :: [KeyHash] -> Gen (Script KeyHash)
+genScript keyHashes = scale (`div` 3) $ sized scriptTree
+    where
+        scriptTree 0 = oneof
+            [ RequireSignatureOf <$> elements keyHashes
+            , ActiveFromSlot <$> arbitrary
+            , ActiveUntilSlot <$> arbitrary
+            ]
+        scriptTree n = do
+            Positive m <- arbitrary
+            let n' = n `div` (m + 1)
+            scripts' <- vectorOf m (scriptTree n')
+            let hasTimelocks = \case
+                    ActiveFromSlot _ -> True
+                    ActiveUntilSlot _ -> True
+                    _ -> False
+            let scriptsWithValidTimelocks = case L.partition hasTimelocks scripts' of
```

Comment:
So the goal is to prune scripts with invalid time locks?
Will this work if there are multiple ActiveFromSlot/ActiveUntilSlot in the generated scripts?
I can think of some nonsensical ranges...
```haskell
let redundant = RequireAnyOf [ActiveFromSlot 2, ActiveFromSlot 5]
    impossible = RequireAllOf [ RequireAnyOf [ActiveFromSlot 5], RequireAnyOf [ActiveFromSlot 2] ]
```

Reply:
this is ok to construct but will not pass validation from cardano-addresses (there is required and recommneded one)